Feature/iconverter

slagroom opened the pull request:
Adds types for aggregating conversions with the same convert-from type as either a map (keyed on convert-to type) or as a converter to perform conversions to dynamically specified convert-to types.

Commit:
Add conversion-map and aggregate-converter types

- Adds interfaces and standard implementations for 2 types:

-- A dictionary-like generic collection that has Type for its key type,
and IConverter<T, U> for its value type. The T (convert-from type) is
the same for all values and is specified by the collection instance's
own type parameter. The U (convert-to type) is specified by a type
parameter on the accessor method.

-- An aggregate converter type that converts from values of T (type
parameter on the instance) to U (type parameter on the convert method).
An implementation is provided that uses the map described above to
provide conversion functions for the type specified by the call.

Both of the types implementations and interfaces have terrible names.
I'm committing them as is in hopes that someone else will learn the
behaviour and come up with a better name than I have.

Commit:
Rename absurdly named types

- Renames the map to 'CoversionMap' and the
aggregate-of-same-from-type-conversions type to Converter. These names
are slighly vague because they don't communicate the fact that the
conversions are grouped by the convert-from type at the instance level,
but they're better than the original working names.

Insomniak47 commented on wimm.Missionary/ConversionMap.cs:
```diff
@@ -0,0 +1,19 @@
+        public IConversion<T, U> Get<U>() =>
+            _conversions.TryGetValue(typeof(U), out var conversion)
+                ? (IConversion<T, U>)conversion
+                : null;
```
How many times do you have to write a maybe type before you start using it :P

slagroom replied:
After I get this lib to 1.0 I'll publish a Secundatives with just the Maybe in it

Insomniak47 commented on wimm.Missionary/IConversionMap.cs:
```diff
@@ -0,0 +1,35 @@
+        /// </returns>
+        IConversion<T, U> Get<U>();
+
+        // TODO: Consider throwing if the same typed conversion is set more than once.
```
I think this is the only reasonable way to do it. I can't imagine why you would need to change this conversion at runtime. It would pooch a lot of the guarantees that you could otherwise make re:construction

slagroom replied:
wf

slagroom replied:
:ship: it?

slagroom replied:
oops left the TODO, wf again

Commit:
Throw on ConversionMap.Set for same U

- Changes the behaviour of the (I)ConversionMap to throw when Set is
called with a <U> type that has already been set.

NOTE: The existing tests for the original set/update behaviour were
actually broken; they just re-tested the simple set/get case again. This
doesn't matter going forward since the behaviour is changed but I wanted
to call myself out before Insomniak47 has the chance.

Commit:
Remove TODO for ConversionMap.Set throw
